pytorch_lightning/utilities/fetching.py

```diff
@@ -174,35 +208,183 @@ class DataFetcher(AbstractDataFetcher):
+    batch 1:                          [HtoD][forward][backward]
+    batch 2:                                                   [HtoD][forward][backward]
+
+    With parallelization, the latency of HtoD copy can be hidden:
+
+    batch 0: [HtoD][forward][backward]
+    batch 1:       [HtoD]             [forward][backward]
+    batch 2:             [HtoD]                          [forward][backward]
+
+    """
+
+    def __init__(
+        self,
+        prefetch_batches: int = 0,
+    ) -> None:
+        super().__init__(prefetch_batches=prefetch_batches)
+
+        self.cuda_stream = torch.cuda.Stream()
+        self.events = []
+
+    @contextlib.contextmanager
+    def fetching_context(self):
+        """Wrap the batch fetching logic under a cuda stream"""
+        with torch.cuda.stream(self.cuda_stream):
+            yield
+
+    def on_fetch_start(self) -> "torch.cuda.Event":
+        # create a cuda event used to record the async stream of data to device.
+        return torch.cuda.Event()
+
+    def on_fetch_end(self, batch, event: torch.cuda.Event) -> None:
+        # move the batch to device and store it
+        super().on_fetch_end(batch)
+
+        # record event and store the event
+        event.record()
+        self.events.append(event)
+
+    def wait(self) -> None:
+        # pop first event from the queue and wait for the batch to be available on device.
+        event = self.events.pop(0)
+        event.wait()
+
+
+class TrainingStepDataLoaderIter:
+
+    """This class is a wrapper to keep track of dataloader iterator fetching event while left entirely to user control."""
+
+    def __init__(self, iterator: Iterator, data_fetcher: "AbstractDataFetcher"):
+        self.iterator = iterator
+        self.data_fetcher = data_fetcher
+
+    def __iter__(self) -> "TrainingStepDataLoaderIter":
+        return self
+
+    def __next__(self) -> Any:
+        self.data_fetcher.fetched += 1
+        try:
+            return next(self.iterator)
+        except StopIteration:
+            self.data_fetcher.done = True
+            raise StopIteration
+
+
+class DataLoaderIterDataFetcher(DataFetcher):
+
+    """
+    This class is used to return directly the `dataloader_iter` to the ``LightningModule`` training_step
+    for users to implement their own pre-fetching logic.
+
+    This feature can be activated as follow:
+
+    Example::
+
+
+        Class MyModel(LightningModule):
+
+            def __init__(self):
+                self.automatic_optimization = False
+
+            def training_step(self, dataloader_iter: Iterator, batch_idx: int) -> None:
+                # it is user responsability to fetch and move the batch to the right device.
+                batch = next(dataloader_iter)
+                batch = batch.to(self.device)
+
+                ...
+
+    """
+
+    def fetching_function(self) -> Generator:
+        iterator = iter(TrainingStepDataLoaderIter(self.dataloader_iter, self))
```

Comment:
this might not be specific to training steps, as overlapping the host to device copy could theoretically provide slight speedups for evaluation & prediction too. cc @yifuwang to check

Reply:
Yes, that's right !